/*
Copyright (c) 2024, Alibaba Group;
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

syntax = "proto3";

import "google/protobuf/empty.proto";

message WarmupResponse {
  bool ok = 1;
}

message WorkerInfo {
  string ip_address = 1;
  int32 instance_id = 2;
  int32 worker_id = 3;
}

message MigrateRequests {
  repeated WorkerInfo src_handlers = 1;
  repeated int32 src_blocks = 2;
  repeated int32 dst_blocks = 3;
}

message SendKvCacheRequest {
  repeated int32 src_blocks = 1;
  repeated int32 dst_blocks = 2;
  optional int32 dst_instance_id = 3;
  optional int32 dst_worker_id = 4;
  optional string key_request_id = 5;
  optional string value_request_id = 6;
  optional string kv_request_id = 7;
}

message SendKvCacheResponse {
  bytes key = 1;
  bytes value = 2;
  bytes kv = 3;
}

service MigrationWorker {
  rpc warmup (google.protobuf.Empty) returns (WarmupResponse) {}
  rpc migrate_cache (MigrateRequests) returns (google.protobuf.Empty) {}
  rpc do_send (SendKvCacheRequest) returns (SendKvCacheResponse) {}
}
